src/matching/Match_patterns.ml

```diff
@@ -37,6 +37,15 @@ let profile_mini_rules = ref false
 (* Types *)
 (*****************************************************************************)
 
+(* Extra context to pass to the environment in the visitor to support implicit return.
+ * More details in Matching_generic.tin.
+ *)
+type imp_ret_ctx = {
```

Comment:
I think it would be better to add this information in AST_generic.stmt instead of in tin.
We could have an extra pass while visiting AST_Generic that populate those fields, and then in Generic_vs_generic
we could just consult those fields; we would not need those imp_ret_enter_xxx because the job would have been done before. Maybe in Naming_AST.ml or in another visitor.

Reply:
@IagoAbal what do you think? @brandonspark ?

Reply:
I would be fine with this, but I am firstly wondering about why we could not insert a `Return` into the AST in the prescribed spots, re my review below.

Reply:
I agree with Yoann here, I would rather keep much of this logic separated. Also, wouldn't be better if we just used the CFG to find out what statements are the "return"ing ones? AST_to_IL probably needs some fixes.

Reply:
> why we could not insert a Return into the AST in the prescribed spots, re my review below.

In case somebody comes across this thread and has the same question, here's [said discussion](https://github.com/returntocorp/semgrep/pull/8551#issuecomment-1696028357).

> Also, wouldn't be better if we just used the CFG to find out what statements are the "return"ing ones? AST_to_IL probably needs some fixes

@IagoAbal I want to make sure I understand you correctly.

You are saying that AST_to_IL has a lot of existing functionality with CFG related stuff. But to find returning statements, AST_to_IL may need a bit more changes.

Is my understanding correct, or did you mean something else?

Reply:
yes but we don't build the IL until we need it in tainting, so it's probably simpler to just extend Naming_AST.ml 
or add another pass in Parse_target.parse_and_resolve_name that do this special tagging of stmts, or
that change some node to add a new ImplReturn AST_generic stmt node.

Reply:
We build the IL already for constant propagation which we do right away after parsing a file, even before any matching is done.

> You are saying that AST_to_IL has a lot of existing functionality with CFG related stuff. But to find returning statements, AST_to_IL may need a bit more changes.

Determining what are the implicit returns should be easy once you have a CFG, I think. The problem is that the construction of the IL may need to be fixed for Ruby/Julia so that that CFG is the correct one. That said, doing that will benefit taint analysis for those languages too.

Since this is something that taint also needs to do (and we also have a limited inter-proc const-prop), I think it's better we do this once in AST_to_IL.